### PR TITLE
Refactor scratch hierarchy 7

### LINKED ISSUE #1235: Refactor out the scratch hierarchy from IBFEMethod
This might be useful to have if we want to scale up `IBMethod` - it would also be nice for writing more classes like IBFEMethod (e.g., the deal.II equivalent). Cleaning up this interface will also make #1226 easier to understand since it will be clearer where the data goes with an explicit interface (instead of IBFEMethod owning everything.

### PR DESCRIPTION
**Pull request by drwells**
Fixes #1235. This patch is longer than the others since it moves both tagging and load balancing into the SecondaryHierarchy (which requires that IBFEMethod no longer do those things in some contexts).

### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

File: ibtk/src/utilities/SecondaryHierarchy.cpp [hunk 1]
```diff
@@ -40,19 +42,100 @@
...
+                TBOX_ASSERT(check_box == finer_box);
+            }
+
+            for (SAMRAI::hier::PatchLevel<NDIM>::Iterator p(level); p; p++)
+            {
+                const Pointer<Patch<NDIM> > patch = level->getPatch(p());
+                const Box<NDIM> box = patch->getBox();
+                Pointer<CellData<NDIM, int> > tag_data = patch->getPatchData(tag_index);
+
+                // SAMRAI's intersection code is wrong and returns nonsense
```
Comment (boyceg) [inline ibtk/src/utilities/SecondaryHierarchy.cpp:113]: Really??

File: ibtk/include/ibtk/SecondaryHierarchy.h [hunk 2]
```diff
@@ -148,6 +137,12 @@ struct SecondaryHierarchy
      */
     int d_finest_patch_level_number;
 
+    /**
+     * Tag strategy. Always does nothing since this class only repartitions and
```
Comment (boyceg) [inline ibtk/include/ibtk/SecondaryHierarchy.h:141]: Can we omit it?

File: ibtk/src/utilities/SecondaryHierarchy.cpp (hunk 1 above)
Comment (drwells) [reply ibtk/src/utilities/SecondaryHierarchy.cpp:113]: Unfortunately yes:
```cpp
template<int DIM> inline
Box<DIM> Box<DIM>::operator*(const Box<DIM>& box) const
{
   Box<DIM> both(*this);
   both.d_lo.max(box.d_lo);
   both.d_hi.min(box.d_hi);
   return(both);
}
```
this produces invalid boxes if the two boxes don't actually intersect. This causes some loops to behave very weirdly inside SAMRAI itself.

Comment (drwells) [reply ibtk/src/utilities/SecondaryHierarchy.cpp:113]: I don't think SAMRAI has a notion of an empty box, but it should create one here.

File: ibtk/include/ibtk/SecondaryHierarchy.h (hunk 2 above)
Comment (drwells) [reply ibtk/include/ibtk/SecondaryHierarchy.h:141]: I should reword this comment - it doesn't really do nothing, it just tags cells to exactly duplicate the tagging done to create the primary hierarchy.

Comment (drwells) [reply ibtk/include/ibtk/SecondaryHierarchy.h]: Fixed!

**Pull request merged**